internal/app/export/service.go

```diff
@@ -4,62 +4,248 @@
+// Start method loads route configuration
+func (e *exporter) Start(queue string) {
 	var route routes.Route
 	for _, r := range e.Cfg.Routes {
 		switch r.Type {
 		case "mfx":
-			route = mfx.NewRoute(r.NatsTopic, r.MqttTopic, r.SubTopic, e.Mqtt, e.Logger)
+			route = mfx.NewRoute(r.NatsTopic, r.MqttTopic, r.SubTopic)
 		default:
-			route = routes.NewRoute(r.NatsTopic, r.MqttTopic, r.SubTopic, e.Mqtt, e.Logger)
+			route = routes.NewRoute(r.NatsTopic, r.MqttTopic, r.SubTopic)
 		}
-		e.Routes = append(e.Routes, route)
-		e.Subscribe(route, queue, nc)
+
+		if e.validateTopic(route.NatsTopic()) {
+			e.Consumers[route.NatsTopic()] = route
+			g, err := e.Cache.GroupCreate(r.NatsTopic, exportGroup)
+			if err != nil {
+				e.Logger.Error(fmt.Sprintf("Failed to create stream group %s", err.Error()))
+			}
+			e.Logger.Info(fmt.Sprintf("Stream group %s created %s", route.NatsTopic(), g))
+		}
+
 	}
+	e.startRepublish()
 }
 
-func (e *exporter) Subscribe(r routes.Route, queue string, nc *nats.Conn) {
-	_, err := nc.QueueSubscribe(r.NatsTopic(), fmt.Sprintf("%s-%s", queue, r.NatsTopic()), r.Consume)
+func (e *exporter) Consume(msg *nats.Msg) {
+	if _, ok := e.Consumers[msg.Subject]; !ok {
+		e.Logger.Info(fmt.Sprintf("no configuration for nats topic %s", msg.Subject))
+		return
+	}
+
+	route := e.Consumers[msg.Subject]
+	payload, err := route.Consume(msg)
 	if err != nil {
-		e.Logger.Error(fmt.Sprintf("Failed to subscribe for NATS/MQTT %s/%s", r.NatsTopic(), r.MqttTopic()))
+		e.Logger.Error(fmt.Sprintf("Failed to consume msg %s", err.Error()))
+	}
+
+	if err = e.publish(route.MqttTopic(), payload); err != nil {
+		_, err := e.Cache.Add(msg.Subject, route.MqttTopic(), payload)
+		if err != nil {
+			e.Logger.Error(fmt.Sprintf("Failed to add to redis stream `%s`", msg.Subject))
+		}
+		return
+	}
+}
+
+func (e *exporter) startRepublish() {
+	// initial connection established on start up
+	<-e.connected
+	e.Logger.Info("Starting republish, waiting for stream data")
+	for _, route := range e.Cfg.Routes {
+		streams := []string{route.NatsTopic, ">"}
+		go func() {
+			for {
+				messages, err := e.Cache.ReadGroup(streams, exportGroup, count, e.ID)
+				if err != nil {
+					e.Logger.Error(fmt.Sprintf("Failed to read from stream %s", err.Error()))
+				}
+				e.Logger.Info(fmt.Sprintf("Read %d records from the stream", len(messages)))
+				e.Logger.Info(fmt.Sprintf("Waiting for connection to %s", e.Cfg.MQTT.Host))
+				// wait for connection
+				for {
+					if e.IsConnected() || <-e.connected {
+						for _, m := range messages {
+							e.publish(m.Topic, []byte(m.Payload))
+						}
+						break
+					}
+				}
+			}
+		}()
+	}
+}
+
+func (e *exporter) Subscribe(topic string, nc *nats.Conn) {
+	_, err := nc.QueueSubscribe(topic, e.ID, e.Consume)
+	if err != nil {
+		e.Logger.Error(fmt.Sprintf("Failed to subscribe for NATS %s", topic))
+	}
+}
+
+func (e *exporter) publish(topic string, payload []byte) error {
+	if e.connectionStatus() == connected {
```

Comment:
Rather exit on not connected, avoid embedding sveral levels of `if`s - this hinders readability.

Reply:
I cannot exit on not connected, not connected is being set in onConnectionLost handler
and if the connection is lost and autoreconnect is set to true when trying to publish and fails it mqtt.publish will not return an error

Reply:
What I am saying instead of checking `if e.connectionStatus() == connected` check `if e.connectionStatus() != connected`, log an error and exit. Otherwise, proceed without else.

Reply:
ah, ok, got it